File: src/main/java/wzq/jcstress/plugin/configuration/JCStressRunConfiguration.java
package wzq.jcstress.plugin.configuration;

import java.util.Collection;
import java.util.HashMap;
import java.util.Map;

import com.intellij.execution.Executor;
import com.intellij.execution.JavaRunConfigurationBase;
import com.intellij.execution.ShortenCommandLine;
import com.intellij.execution.configurations.ConfigurationFactory;
import com.intellij.execution.configurations.JavaRunConfigurationModule;
import com.intellij.execution.configurations.RunConfiguration;
import com.intellij.execution.configurations.RunProfileState;
import com.intellij.execution.runners.ExecutionEnvironment;
import com.intellij.openapi.module.Module;
import com.intellij.openapi.options.SettingsEditor;
import com.intellij.openapi.util.InvalidDataException;
import com.intellij.openapi.util.WriteExternalException;
import org.jdom.Element;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * @author 吴志强
 * @date 2022/8/24
 */
public class JCStressRunConfiguration extends JavaRunConfigurationBase {

    private String vmParameters;
    private String programParameters;
    private String workingDirectory;
    private Map<String, String> envs = new HashMap<>();
    private boolean passParentEnvs;
    private ShortenCommandLine shortenCommandLine;
    private boolean alternativeJrePathEnabled;
    private String alternativeJrePath;
    private String jcstressClass;

    public JCStressRunConfiguration(
            String name,
            @NotNull JavaRunConfigurationModule configurationModule,
            @NotNull ConfigurationFactory factory) {
        super(name, configurationModule, factory);
    }

    @Override
    public void setVMParameters(@Nullable String s) {
        this.vmParameters = s;
    }

    @Override
    public String getVMParameters() {
        return this.vmParameters;
    }

    @Override
    public boolean isAlternativeJrePathEnabled() {
        return this.alternativeJrePathEnabled;
    }

    @Override
    public void setAlternativeJrePathEnabled(boolean b) {
        this.alternativeJrePathEnabled = b;
    }

    @Override
    public @Nullable String getAlternativeJrePath() {
        return this.alternativeJrePath;
    }

    @Override
    public void setAlternativeJrePath(@Nullable String path) {
        this.alternativeJrePath = path;
    }

    @Override
    public @Nullable String getRunClass() {
        return "org.openjdk.jcstress.Main";
    }

    @Override
    public @Nullable String getPackage() {
        return null;
    }

    @Override
    public void setProgramParameters(@Nullable String s) {
        this.programParameters = s;
    }

    @Override
    public @Nullable String getProgramParameters() {
        return this.programParameters;
    }

    @Override
    public void setWorkingDirectory(@Nullable String s) {
        this.workingDirectory = s;
    }

    @Override
    public @Nullable String getWorkingDirectory() {
        return this.workingDirectory;
    }

    @Override
    public void setEnvs(@NotNull Map<String, String> map) {
        this.envs = map;
    }

    @Override
    public @NotNull Map<String, String> getEnvs() {
        return this.envs;
    }

    @Override
    public void setPassParentEnvs(boolean b) {
        this.passParentEnvs = b;
    }

    @Override
    public boolean isPassParentEnvs() {
        return this.passParentEnvs;
    }

    @Override
    public @Nullable ShortenCommandLine getShortenCommandLine() {
        return this.shortenCommandLine;
    }

    @Override
    public void setShortenCommandLine(@Nullable ShortenCommandLine shortenCommandLine) {
        this.shortenCommandLine = shortenCommandLine;
    }

    public String getJCStressClass() {
        return jcstressClass;
    }

    public void setJCStressClass(String jcstressClass) {
        this.jcstressClass = jcstressClass;
    }

    @Override
    public Collection<Module> getValidModules() {
        return null;
    }

    @Override
    public @NotNull SettingsEditor<? extends RunConfiguration> getConfigurationEditor() {
        return new JCStressSettingsEditor();
    }

    @Override
    public @Nullable RunProfileState getState(
            @NotNull Executor executor, @NotNull ExecutionEnvironment executionEnvironment) {
        return new JCStressCommandLineState(executionEnvironment, this);
    }

    @Override
    public void readExternal(@NotNull Element element) throws InvalidDataException {
        super.readExternal(element);
        this.vmParameters = element.getAttributeValue("vmParameters");
        this.programParameters = element.getAttributeValue("programParameters");
        this.workingDirectory = element.getAttributeValue("workingDirectory");
        this.jcstressClass = element.getAttributeValue("jcstressClass");
    }

    @Override
    public void writeExternal(@NotNull Element element) throws WriteExternalException {
        super.writeExternal(element);
        if (this.vmParameters != null) {
            element.setAttribute("vmParameters", this.vmParameters);
        }
        if (this.programParameters != null) {
            element.setAttribute("programParameters", this.programParameters);
        }
        if (this.workingDirectory != null) {
            element.setAttribute("workingDirectory", this.workingDirectory);
        }
        if (this.jcstressClass != null) {
            element.setAttribute("jcstressClass", this.jcstressClass);
        }
    }
}

File: src/main/kotlin/wzq/jcstress/plugin/JCStressRunConfiguration.kt
package wzq.jcstress.plugin

import com.intellij.execution.Executor
import com.intellij.execution.JavaRunConfigurationBase
import com.intellij.execution.ShortenCommandLine
import com.intellij.execution.configurations.ConfigurationFactory
import com.intellij.execution.configurations.JavaRunConfigurationModule
import com.intellij.execution.configurations.RunConfiguration
import com.intellij.execution.configurations.RunProfileState
import com.intellij.execution.runners.ExecutionEnvironment
import com.intellij.openapi.module.Module
import com.intellij.openapi.options.SettingsEditor
import org.jdom.Element

/**
 * @author 吴志强
 */
class JCStressRunConfiguration(
    name: String,
    javaRunConfigurationModule: JavaRunConfigurationModule,
    configurationFactory: ConfigurationFactory,
) : JavaRunConfigurationBase(name, javaRunConfigurationModule, configurationFactory) {
    private var vmParameters: String? = null
    private var alternativeJrePathEnabled = false
    private var alternativeJrePath: String? = null
    private var programParameters: String? = null
    private var workingDirectory: String? = null
    private var envs: Map<String?, String?> = hashMapOf()
    private var passParentEnvs = false
    private var shortenCommandLine: ShortenCommandLine? = null
    var jcstressClass: String? = null

    override fun getValidModules(): Collection<Module?>? = null

    override fun getState(
        executor: Executor,
        environment: ExecutionEnvironment,
    ): RunProfileState? = JCStressCommandLineState(environment, this)

    override fun getConfigurationEditor(): SettingsEditor<out RunConfiguration?> = JCStressSettingsEditor()

    override fun setVMParameters(vmParameters: String?) {
        this.vmParameters = vmParameters
    }

    override fun getVMParameters(): String? = this.vmParameters

    override fun isAlternativeJrePathEnabled(): Boolean = this.alternativeJrePathEnabled

    override fun setAlternativeJrePathEnabled(alternativeJrePathEnabled: Boolean) {
        this.alternativeJrePathEnabled = alternativeJrePathEnabled
    }

    override fun getAlternativeJrePath(): String? = this.alternativeJrePath

    override fun setAlternativeJrePath(alternativeJrePath: String?) {
        this.alternativeJrePath = alternativeJrePath
    }

    override fun getRunClass(): String? = "org.openjdk.jcstress.Main"

    override fun getPackage(): String? = null

    override fun setProgramParameters(programParameters: String?) {
        this.programParameters = programParameters
    }

    override fun getProgramParameters(): String? = this.programParameters

    override fun setWorkingDirectory(workingDirectory: String?) {
        this.workingDirectory = workingDirectory
    }

    override fun getWorkingDirectory(): String? = this.workingDirectory

    override fun setEnvs(envs: Map<String?, String?>) {
        this.envs = envs
    }

    override fun getEnvs(): Map<String?, String?> = this.envs

    override fun setPassParentEnvs(passParentEnvs: Boolean) {
        this.passParentEnvs = passParentEnvs
    }

    override fun isPassParentEnvs(): Boolean = this.passParentEnvs

    override fun getShortenCommandLine(): ShortenCommandLine? = this.shortenCommandLine

    override fun setShortenCommandLine(shortenCommandLine: ShortenCommandLine?) {
        this.shortenCommandLine = shortenCommandLine
    }

    override fun readExternal(element: Element) {
        super.readExternal(element)
        this.vmParameters = element.getAttributeValue("vmParameters")
        this.programParameters = element.getAttributeValue("programParameters")
        this.workingDirectory = element.getAttributeValue("workingDirectory")
        this.jcstressClass = element.getAttributeValue("jcstressClass")
    }

    override fun writeExternal(element: Element) {
        super.writeExternal(element)
        if (this.vmParameters != null) {
            element.setAttribute("vmParameters", this.vmParameters)
        }
        if (this.programParameters != null) {
            element.setAttribute("programParameters", this.programParameters)
        }
        if (this.workingDirectory != null) {
            element.setAttribute("workingDirectory", this.workingDirectory)
        }
        if (this.jcstressClass != null) {
            element.setAttribute("jcstressClass", this.jcstressClass)
        }
    }
}
